pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -465,17 +467,22 @@ public CompletableFuture<Void> purgeTxns(List<Long> dataLedgers) {
 
     @Override
     public CompletableFuture<Void> clearSnapshot() {
-        return this.takeSnapshotWriter.thenCompose(writer -> {
+        return this.takeSnapshotWriter.getFuture().thenCompose(writer -> {
             TransactionBufferSnapshot snapshot = new TransactionBufferSnapshot();
             snapshot.setTopicName(topic.getName());
             return writer.deleteAsync(snapshot);
         }).thenCompose(__ -> CompletableFuture.completedFuture(null));
     }
 
     @Override
-    public CompletableFuture<Void> closeAsync() {
+    public synchronized CompletableFuture<Void> closeAsync() {
```

Comment:
```
    @Override
    public synchronized CompletableFuture<Void> closeAsync() {
        if (!checkIfClose()) {
            // make sure the snapshot counted writer only release one time.
            this.topic.getBrokerService().getPulsar().getTransactionBufferSnapshotService()
                    .releaseReferenceWriter(this.takeSnapshotWriter);
        }
        changeToCloseState();
        return CompletableFuture.completedFuture(null);
    }
```

I'm afraid this method is not a thread-safe operation. If multi-threads call this method (maybe this will not happen, but it's possible) the release operation will execute multi times.